src/sentry/filters/preprocess_hashes.py

```diff
@@ -11,3 +17,42 @@
 
 def get_raw_cache_key(project_id, event_id):
     return 'e:raw:{1}:{0}'.format(project_id, event_id)
+
+
+def get_preprocess_hash_inputs(event):
+    return get_preprocess_hash_inputs_with_reason(event)[1]
+
+
+def get_preprocess_hash_inputs_with_reason(data):
+    interfaces = get_interfaces(data)
+    for interface in six.itervalues(interfaces):
+        # normalize_in_app hasn't run on the data, so
+        # `in_app` isn't necessarily accurate
+        result = interface.get_hash(data['platform'], system_frames=True, is_processed_data=False)
+        if not result:
+            continue
+        return (interface.get_path(), [result])
+    return ('message', [data['message']])
```

Comment:
I guess we do this in `get_hashes_for_event_with_reason` but assuming that `data['message']` exists makes me a little nervous.

Reply:
are you worried because you think it's more likely that that would happen before processing? if so, i can change it to be more cautious and do something like raise an error that i can then handle wherever this is used

Reply:
Yeah, I don't really think there is any enforcement upstream that this key would actually be set, is there? (… and, if there is incidentally, do you trust that somebody isn't going to change it and cause this to break?)

Aside, I honestly don't even understand how this works in the existing event manager case, since `get_hashes_for_event` is called after the `message` attribute has been popped from the data dictionary:

https://github.com/getsentry/sentry/blob/486fd31e6ed58df6b25f96119238ba93a8615c8e/src/sentry/event_manager.py#L427

At the point where `get_hashes_for_event` could be called, it hasn't been set back on the event yet  (`event.message` is actually [set later here](https://github.com/getsentry/sentry/blob/486fd31e6ed58df6b25f96119238ba93a8615c8e/src/sentry/event_manager.py#L566)): https://github.com/getsentry/sentry/blob/486fd31e6ed58df6b25f96119238ba93a8615c8e/src/sentry/event_manager.py#L528

I guess this must just be returning the hash of the empty string, like this?

```python
>>> e = Event()
>>> get_hashes_for_event_with_reason(e)
('message', [u''])
```

Reply:
yeah makes sense, and in digging into this more, i think in my case, it would be possible to hit a KeyError, so I will definitely fix.
 
As far as your question about how this is working in the existing `get_hashes_for_event`: from what i can tell, for the message attribute to be correctly used, `EventManager.normalize` needs to be called before `save`. `normalize` [adds the message](https://github.com/getsentry/sentry/blob/486fd31e6ed58df6b25f96119238ba93a8615c8e/src/sentry/event_manager.py#L350) interface, which allows the message to be used for grouping if provided. So I think you're right and what that probably means is that when we hit `return ('message', [event.message])`, it will never return anything except `('message', [u''])` even if a message was provided. Should I fix that in some way? Even just making that explicit could be helpful maybe

Reply:
Hm, I'm honestly unsure what the right move would be here. Erroring out doesn't seem an option since we see these in the wild — mainly from events that either didn't have any hashable interfaces in the first place (not sure why people are sending these) or events where all hashable interfaces were discarded (either invalid user provided data, or SDK errors.)

I think returning an empty string (and something other than "message" as the reason) would be clearer here, at least, with a comment explaining what's going on (so that we don't have to dig through this again later to come to the same conclusion, as you mentioned on Slack.)